File: analytics/src/test/java/io/appmetrica/analytics/impl/AppMetricaImplFieldsProviderTest.java
package io.appmetrica.analytics.impl;

import android.content.Context;
import android.os.Handler;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import io.appmetrica.analytics.impl.client.ProcessConfiguration;
import io.appmetrica.analytics.impl.db.preferences.PreferencesClientDbStorage;
import io.appmetrica.analytics.impl.referrer.client.ReferrerHelper;
import io.appmetrica.analytics.impl.startup.StartupHelper;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.ConstructionArgumentCaptor;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockedConstruction;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(RobolectricTestRunner.class)
public class AppMetricaImplFieldsProviderTest extends CommonTest {

    @Mock
    private Handler mHandler;
    @Mock
    private AppMetricaImpl mAppMetrica;
    @Mock
    private DataResultReceiver mReceiver;
    @Mock
    private ProcessConfiguration mProcessConfiguration;
    @Mock
    private ReportsHandler mReportsHandler;
    @Mock
    private ICommonExecutor mCommonExecutor;
    @Mock
    private StartupHelper mStartupHelper;
    @Mock
    private PreferencesClientDbStorage mPreferences;
    private Context mContext;
    private AppMetricaImplFieldsProvider mFieldsProvider;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        mContext = RuntimeEnvironment.getApplication();
        mFieldsProvider = new AppMetricaImplFieldsProvider();
    }

    @Test
    public void createDataResultReceiver() {
        ConstructionArgumentCaptor<DataResultReceiver> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<DataResultReceiver> constr = Mockito.mockConstruction(
                DataResultReceiver.class, interceptor
        )) {
            mFieldsProvider.createDataResultReceiver(mHandler, mAppMetrica);
            assertThat(constr.constructed()).hasSize(1);

            assertThat(interceptor.flatArguments()).containsExactly(mHandler, mAppMetrica);
        }
    }

    @Test
    public void createProcessConfiguration() {
        ConstructionArgumentCaptor<ProcessConfiguration> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<ProcessConfiguration> constr = Mockito.mockConstruction(
                ProcessConfiguration.class, interceptor
        )) {
            mFieldsProvider.createProcessConfiguration(mContext, mReceiver);
            assertThat(constr.constructed()).hasSize(1);
            assertThat(interceptor.flatArguments()).containsExactly(mContext, mReceiver);
        }
    }

    @Test
    public void createReportsHandler() {
        ConstructionArgumentCaptor<ReportsHandler> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<ReportsHandler> constr = Mockito.mockConstruction(
                ReportsHandler.class, interceptor
        )) {
            mFieldsProvider.createReportsHandler(mProcessConfiguration, mContext, mCommonExecutor);
            assertThat(constr.constructed()).hasSize(1);
            assertThat(interceptor.flatArguments()).containsExactly(mProcessConfiguration, mContext, mCommonExecutor);
        }
    }

    @Test
    public void createStartupHelper() {
        ConstructionArgumentCaptor<StartupHelper> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<StartupHelper> constr = Mockito.mockConstruction(
                StartupHelper.class, interceptor
        )) {
            mFieldsProvider.createStartupHelper(mContext, mReportsHandler, mPreferences, mHandler);
            assertThat(constr.constructed()).hasSize(1);
            assertThat(interceptor.flatArguments()).containsExactly(mContext, mReportsHandler, mPreferences, mHandler);
        }
    }

    @Test
    public void createReferrerHelper() {
        ConstructionArgumentCaptor<ReferrerHelper> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<ReferrerHelper> constr = Mockito.mockConstruction(
                ReferrerHelper.class, interceptor
        )) {
            mFieldsProvider.createReferrerHelper(mReportsHandler, mPreferences, mHandler);
            assertThat(constr.constructed()).hasSize(1);
            assertThat(interceptor.flatArguments()).containsExactly(mReportsHandler, mPreferences, mHandler);
        }
    }

    @Test
    public void createReporterFactory() {
        ConstructionArgumentCaptor<ReporterFactory> interceptor = new ConstructionArgumentCaptor<>();
        try (MockedConstruction<ReporterFactory> constr = Mockito.mockConstruction(
                ReporterFactory.class, interceptor
        )) {
            mFieldsProvider.createReporterFactory(mContext, mProcessConfiguration, mReportsHandler, mHandler, mStartupHelper);
            assertThat(constr.constructed()).hasSize(1);
            assertThat(interceptor.flatArguments()).containsExactly(mContext, mProcessConfiguration, mReportsHandler, mHandler, mStartupHelper);
        }
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/AppMetricaImplFieldsProviderTest.kt
package io.appmetrica.analytics.impl

import android.content.Context
import android.os.Handler
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor
import io.appmetrica.analytics.impl.client.ProcessConfiguration
import io.appmetrica.analytics.impl.db.preferences.PreferencesClientDbStorage
import io.appmetrica.analytics.impl.modules.ModuleStatusReporter
import io.appmetrica.analytics.impl.referrer.client.ReferrerHelper
import io.appmetrica.analytics.impl.startup.StartupHelper
import io.appmetrica.analytics.testutils.ClientServiceLocatorRule
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.constructionRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
internal class AppMetricaImplFieldsProviderTest : CommonTest() {

    private val handler: Handler = mock()
    private val appMetrica: AppMetricaImpl = mock()
    private val receiver: DataResultReceiver = mock()
    private val processConfiguration: ProcessConfiguration = mock()
    private val reportsHandler: ReportsHandler = mock()
    private val commonExecutor: ICommonExecutor = mock()
    private val startupHelper: StartupHelper = mock()
    private val preferences: PreferencesClientDbStorage = mock()
    private var context: Context = mock()

    @get:Rule
    val clientServiceLocatorRule = ClientServiceLocatorRule()

    @get:Rule
    val dataResultReceiverRule = constructionRule<DataResultReceiver>()

    @get:Rule
    val processConfigurationRule = constructionRule<ProcessConfiguration>()

    @get:Rule
    val reportsHandlerRule = constructionRule<ReportsHandler>()

    @get:Rule
    val startupHelperRule = constructionRule<StartupHelper>()

    @get:Rule
    val referrerHelperRule = constructionRule<ReferrerHelper>()

    @get:Rule
    val reporterFactoryRule = constructionRule<ReporterFactory>()

    @get:Rule
    val moduleStatusReporterRule = constructionRule<ModuleStatusReporter>()

    private val fieldsProvider by setUp { AppMetricaImplFieldsProvider() }

    @Test
    fun createDataResultReceiver() {
        assertThat(fieldsProvider.createDataResultReceiver(handler, appMetrica))
            .isEqualTo(dataResultReceiverRule.constructionMock.constructed().first())
        assertThat(dataResultReceiverRule.constructionMock.constructed()).hasSize(1)
        assertThat(dataResultReceiverRule.argumentInterceptor.flatArguments())
            .containsExactly(handler, appMetrica)
    }

    @Test
    fun createProcessConfiguration() {
        assertThat(fieldsProvider.createProcessConfiguration(context, receiver))
            .isEqualTo(processConfigurationRule.constructionMock.constructed().first())
        assertThat(processConfigurationRule.constructionMock.constructed())
            .hasSize(1)
        assertThat(processConfigurationRule.argumentInterceptor.flatArguments())
            .containsExactly(context, receiver)
    }

    @Test
    fun createReportsHandler() {
        assertThat(fieldsProvider.createReportsHandler(processConfiguration, context, commonExecutor))
            .isEqualTo(reportsHandlerRule.constructionMock.constructed().first())
        assertThat(reportsHandlerRule.constructionMock.constructed()).hasSize(1)
        assertThat(reportsHandlerRule.argumentInterceptor.flatArguments())
            .containsExactly(processConfiguration, context, commonExecutor)
    }

    @Test
    fun createStartupHelper() {
        assertThat(fieldsProvider.createStartupHelper(context, reportsHandler, handler))
            .isEqualTo(startupHelperRule.constructionMock.constructed().first())
        assertThat(startupHelperRule.constructionMock.constructed()).hasSize(1)
        assertThat(startupHelperRule.argumentInterceptor.flatArguments())
            .containsExactly(reportsHandler, ClientServiceLocator.getInstance().getStartupParams(context), handler)
    }

    @Test
    fun createReferrerHelper() {
        assertThat(fieldsProvider.createReferrerHelper(reportsHandler, preferences, handler))
            .isEqualTo(referrerHelperRule.constructionMock.constructed().first())
        assertThat(referrerHelperRule.constructionMock.constructed()).hasSize(1)
        assertThat(referrerHelperRule.argumentInterceptor.flatArguments())
            .containsExactly(reportsHandler, preferences, handler)
    }

    @Test
    fun createReporterFactory() {
        val reporterFactory = fieldsProvider.createReporterFactory(
            context,
            processConfiguration,
            reportsHandler,
            handler,
            startupHelper
        )
        assertThat(reporterFactory).isEqualTo(reporterFactoryRule.constructionMock.constructed().first())
        assertThat(reporterFactoryRule.constructionMock.constructed()).hasSize(1)
        assertThat(reporterFactoryRule.argumentInterceptor.flatArguments())
            .containsExactly(
                context,
                processConfiguration,
                reportsHandler,
                handler,
                startupHelper
            )
    }

    @Test
    fun createModuleStatusReporter() {
        val moduleStatusReporter = fieldsProvider.createModuleStatusReporter(context)
        assertThat(moduleStatusReporter).isEqualTo(moduleStatusReporterRule.constructionMock.constructed().first())
        assertThat(moduleStatusReporterRule.constructionMock.constructed()).hasSize(1)
        assertThat(moduleStatusReporterRule.argumentInterceptor.flatArguments())
            .contains(
                ClientServiceLocator.getInstance().clientExecutorProvider.defaultExecutor,
                ClientServiceLocator.getInstance().getPreferencesClientDbStorage(context),
                "client_modules"
            )
    }
}
